website/src/app/blog/sans-io/readme.mdx

```diff
@@ -0,0 +1,556 @@
+In a nutshell, it discusses the constraint that async functions can only be called from other async functions, thus "colouring" them.
+There are various takes on this but what stands out for me is that the ability to suspend execution and resume later is a pretty important part of function's API contract.
+The fact that Rust enforces this at compile-time is a good thing.
+
+A result of this constraint is that an async function deep down in your stack "forces" every calling function to also become async in order to `.await` the inner function.
+This can be problematic if the code you want to call isn't actually yours but a dependency that you are pulling in.
+
+Some people see this as a problem and they would like to write code that is agnostic over the "asyncness" of their dependencies.
+This idea has merit.
+At the very bottom of each async call stack sits a `Future` that needs to suspend on something.
+Usually, this is some form of IO, like writing to a socket, reading from a file, waiting for time to advance, etc.
+The majority of async functions don't actually perform async work themselves.
+Instead, they depend on an async primitive like a `tokio::net::UdpSocket`.
+The code around those inner async functions would usually also work in a blocking context, but the author of your dependency happened to pick the async variant.
+
+Firezone's connectivity library `connlib` uses ICE and as part of that, we utilise STUN to discover our server-reflexive candidate, i.e. our public address.
+STUN is a binary message format and a STUN binding is a pretty simple protocol: Send a UDP packet to server, server notes the IP + port it sees as the sending socket and send a UDP packet back containing that address.
+
+Here is how we could implement this using `tokio`'s `UdpSocket` (thank you to Cloudflare for the public STUN server):
+
+```rust
+#[tokio::main]
+async fn main() -> anyhow::Result<()> {
+    let socket = UdpSocket::bind("0.0.0.0:0").await?;
+    socket.connect("stun.cloudflare.com:3478").await?;
+    socket.send(&make_binding_request()).await?;
+
+    let mut buf = vec![0u8; 100];
+    let num_read = socket.recv(&mut buf).await?;
+    let address = parse_binding_response(&buf[..num_read]);
+
+    println!("Our public IP is: {address}");
+
+    Ok(())
+}
+```
+
+This could be also be written using blocking IO from the standard library:
+
+```rust
+fn main() -> anyhow::Result<()> {
+    let socket = UdpSocket::bind("0.0.0.0:0")?;
+    socket.connect("stun.cloudflare.com:3478")?;
+    socket.send(&make_binding_request())?;
+
+    let mut buf = vec![0u8; 100];
+    let num_read = socket.recv(&mut buf)?;
+    let address = parse_binding_response(&buf[..num_read]);
+
+    println!("Our public IP is: {address}");
+
+    Ok(())
+}
+```
+
+You can find all of these snippets as working programs in the following repository: https://github.com/firezone/sans-io-blog-example.
```

Comment:
I have the repo ready locally but don't have permission to make one.

Reply:
I'll create it real quick.

Reply:
https://github.com/firezone/sans-io-blog-example